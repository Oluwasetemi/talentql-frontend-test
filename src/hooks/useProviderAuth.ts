import React from 'react';
import { fakeAuth } from '../App';

export function useProvideAuth() {
	const [user, setUser] = React.useState<string | null>(null);

	const signin = (cb: () => void) => {
		return fakeAuth.signin(() => {
			setUser('user');
			cb();
		});
	};

	const signout = (cb: () => void) => {
		return fakeAuth.signout(() => {
			setUser(null);
			cb();
		});
	};

	return {
		user,
		signin,
		signout,
	};
}
