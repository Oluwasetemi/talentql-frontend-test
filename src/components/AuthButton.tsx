import * as React from 'react';
import { useHistory } from 'react-router-dom';
import { useAuth } from '../hooks/useAuth';

export function AuthButton() {
	let history = useHistory();
	let auth = useAuth();

	return auth!.user ? (
		<p>
			Welcome!{' '}
			<button
				data-testid="signout"
				onClick={() => {
					auth!.signout(() => history.push('/'));
				}}
			>
				Sign out
			</button>
		</p>
	) : (
		<p>You are not logged in.</p>
	);
}
