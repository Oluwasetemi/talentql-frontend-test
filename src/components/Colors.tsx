import { useToasts } from '@keystone-ui/toast';
import React from 'react';
import styled from 'styled-components';
import { checkColor, resetColors, unCheckColor } from '../store/color';
import { useAppDispatch, useAppSelector } from '../store/hooks';

export const ColorContainer = styled.div`
	/* height: 50px; */
	display: flex;
	gap: 10px;
	flex-wrap: wrap;
`;
export const ColorItem = styled.div`
	width: 50px;
	height: 50px;
	border-radius: 50%;
	position: relative;
	background-color: ${(prop) => prop.color};
	border: 1px solid transparent;

	input[type='checkbox'] {
		visibility: hidden;
	}

	label {
		background-color: #fff;
		border: 1px solid #ccc;
		border-radius: 50%;
		cursor: pointer;
		height: 50px;
		left: 0;
		position: absolute;
		top: 0;
		width: 50px;
	}

	input[type='checkbox']:checked + label {
		box-shadow: 0 0 5px rgba(81, 203, 238, 1);
		border: 1px solid rgba(81, 203, 238, 1);
	}

	input[type='checkbox']:checked + label:after {
		opacity: 1;
	}
`;

export function ColorsFilter() {
	const { addToast } = useToasts();
	const colors = useAppSelector((state) => state.color.colors).map(
		(color) => color.color,
	);
	const selectedColors = useAppSelector((state) => state.color.selectedColors);

	const dispatch = useAppDispatch();

	function isChecked(color: string) {
		return selectedColors.includes(color);
	}

	const handleOnChange = (event: React.ChangeEvent<HTMLInputElement>) => {
		const name = event.target.name;
		// eslint-disable-next-line no-unused-expressions
		if (isChecked(name)) {
			dispatch(unCheckColor(name));
			addToast({ title: `${name} color unchecked`, tone: 'positive' });
		} else {
			dispatch(checkColor(name));
			addToast({ title: `${name} color checked`, tone: 'negative' });
		}
	};

	React.useEffect(() => {
		if (selectedColors.length === 1) {
			let current = document.querySelector('div.current') as HTMLElement;
			current.addEventListener(
				'click',
				() => {
					setTimeout(() => {
						dispatch(resetColors());
					}, 0);
				},
				{ once: true },
			);
		}
	}, [selectedColors]);

	return (
		<ColorContainer>
			{colors.map((color, index) => (
				<ColorItem
					key={index}
					className={isChecked(color) ? 'current' : ''}
					color={color}
					data-testid={color}
				>
					<input
						data-testid="color-checkbox"
						type="checkbox"
						name={color}
						id={color}
						checked={isChecked(color)}
						onChange={handleOnChange}
					/>
					<label htmlFor={color} style={{ backgroundColor: `${color}` }} />
				</ColorItem>
			))}
		</ColorContainer>
	);
}
