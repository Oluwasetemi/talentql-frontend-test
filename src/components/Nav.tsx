import * as React from 'react';
import { useHistory } from 'react-router-dom';
import styled from 'styled-components';
import { useAuth } from '../hooks/useAuth';
import logout from '../logout.svg';

export const NavBar = styled.header`
	background: #fff;
	width: 85%;
	margin: 0 auto;
	display: flex;
	justify-content: space-between;
	align-items: center;
	height: 45px;
	span {
		display: flex;
	}
	.logo {
		text-transform: uppercase;
	}
	span img {
		/* margin-bottom: 10px; */
		height: 20px;
		align-content: center;
		margin-left: 10px;
	}
	.logout {
		color: red;
	}
`;

export function Nav() {
	let history = useHistory();
	let auth = useAuth();
	let { from } = { from: { pathname: '/' } };

	return (
		<NavBar>
			<span className="logo">Shapes</span>
			{auth!.user ? (
				<span
					onClick={() => {
						auth!.signout(() => history.push('/'));
					}}
					className="logout"
					data-testid="logout"
				>
					logout <img src={logout} alt="logout" />
				</span>
			) : (
				<span
					onClick={() => {
						auth!.signin(() => history.replace(from));
					}}
					className="logout"
					data-testid="signin"
					role="button"
				>
					login
				</span>
			)}
		</NavBar>
	);
}
