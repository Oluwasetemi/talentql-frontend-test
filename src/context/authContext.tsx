import React from 'react';
import { useProvideAuth } from '../hooks/useProviderAuth';

export interface AppContextInterface {
	user: string | null;
	signin: (cb: () => void) => void;
	signout: (cb: () => void) => void;
}

export const authContext = React.createContext<AppContextInterface | null>(
	null,
);

export function ProvideAuth({ children }: { children: React.ReactNode }) {
	const auth = useProvideAuth();
	return <authContext.Provider value={auth}>{children}</authContext.Provider>;
}
