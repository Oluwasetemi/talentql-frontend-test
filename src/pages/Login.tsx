import * as React from 'react';
import { useHistory, useLocation } from 'react-router-dom';
import { AuthButton } from '../components/AuthButton';
import { Navigation } from '../components/Navigation';
import { useAuth } from '../hooks/useAuth';

export function Login(): JSX.Element {
	let history = useHistory();
	let location = useLocation();
	let auth = useAuth();

	let { from } = (location.state as any) || { from: { pathname: '/' } };

	let login = () => {
		auth!.signin(() => {
			history.replace(from);
		});
	};

	return (
		<Navigation>
			<AuthButton />
			<h3>Login</h3>
			<p>You must log in to view the Color and Shape filters</p>
			<button data-testid="login" onClick={login}>
				Log in
			</button>
		</Navigation>
	);
}
