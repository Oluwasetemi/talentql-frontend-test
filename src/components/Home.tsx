import * as React from 'react';
import { AppWrapper } from '../App';
import { AuthButton } from './AuthButton';
import { ColorsFilter } from './Colors';
import { Nav } from './Nav';
import { RenderBox } from './RenderBox';
import { ShapesFilter } from './Shapes';

function ColorDashboard() {
	return (
		<>
			<Nav />
			<div className="App-body">
				<AppWrapper>
					<AuthButton />
					<h3>Filters</h3>
					<h5 className="blue-text">Shapes</h5>
					<ShapesFilter />
					<h5 className="blue-text">Color.colors</h5>
					<ColorsFilter />
					<RenderBox />
				</AppWrapper>
			</div>
		</>
	);
}

export default ColorDashboard;
