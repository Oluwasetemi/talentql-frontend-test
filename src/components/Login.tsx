import * as React from 'react';
import { useHistory, useLocation } from 'react-router-dom';
import { AppWrapper } from '../App';
import { useAuth } from '../hooks/useAuth';
import { AuthButton } from './AuthButton';
import { Nav } from './Nav';

export function Login(): JSX.Element {
  let history = useHistory();
  let location = useLocation();
  let auth = useAuth();

  let { from } = location.state as any || { from: { pathname: '/' } };
  let login = () => {
    auth!.signin(() => {
      history.replace(from);
    });
  };
  return (
    <>
      <Nav />
      <div className="App-body">
        <AppWrapper>
          <AuthButton />
          <h3>Login</h3>
          <p>You must log in to view the page at {from.pathname}</p>
          <button onClick={login}>Log in</button>
        </AppWrapper>
      </div>
    </>
  );
}