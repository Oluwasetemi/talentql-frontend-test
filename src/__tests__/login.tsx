import { fireEvent, render } from '@testing-library/react';
import 'jest-styled-components';
import * as React from 'react';
import { Login } from '../components/Login';
import { authContext } from '../context/authContext';
import { useProvideAuth } from '../hooks/useProviderAuth';

const mockUseLocationValue = {
	pathname: '/login',
	search: '',
	hash: '',
	state: null,
};

jest.mock('react-router-dom', () => ({
	...(jest.requireActual('react-router-dom') as {}),
	useLocation: jest.fn().mockImplementation(() => {
		return mockUseLocationValue;
	}),
}));

function LoginApp() {
	const auth = useProvideAuth();
	return (
		<authContext.Provider value={auth}>
			<Login />
		</authContext.Provider>
	);
}

test('login component has a button log in', () => {
	const { getAllByText } = render(<LoginApp />);
	const input = getAllByText(/log in/i);
	expect(input[1]).toMatchSnapshot();
});

test('fire button login in the has a button', () => {
	const { getAllByText } = render(<LoginApp />);
	const input = getAllByText(/log in/i);
	fireEvent.click(input[1]);
});
