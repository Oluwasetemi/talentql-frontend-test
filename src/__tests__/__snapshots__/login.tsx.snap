// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login component has a button log in 1`] = `
<button
  data-testid="login"
>
  Log in
</button>
`;
